### DISCUSSION TITLE
Particles

### DISCUSSION THREAD

**Discussion by Ono-Sendai** [jrouwe/JoltPhysics#756] [General]
Hi Jorrit,
As part of adding some vehicle effects to Substrata, I want to add some particles.  For example, pieces of dirt and dust flung up from tire contact with the ground from a motorbike, or water spray from a boat.

Do you have any suggestions for how to do that in Jolt?  Are you planning any kind of particle support?

One possibility is that I do all particles on the GPU, bypassing Jolt, but I thought I would start with CPU-based physics particles and see how the performance is.

I'll probably just start with using low mass bodies with simple shapes (sphere / capsule/ box?) for particles.  However this still may be too heavyweight in terms of performance. 
One of the defining properties of particles, I think, is that they have negligible momentum, so we don't need to run a full collision and momentum transfer computation when they hit something.
The other defining property would be their small size (point-like).  Maybe this can be used for some efficiency gain.
An additional property would be their similarity with each other.  This would lend them to being allocated or computed in some kind of batched way.

Anyway what do you think? Custom particle support in Jolt?
Will try with existing capabilities in the mean time anyway.

**Comment by jrouwe** [comment]
I was assuming that every engine had their own particle system for VFX. Usually there's a lot of smoke and mirrors involved in them, so I'm not sure if there's much value in adding something like that to a physics engine.

That said, technically Jolt already has a particle system in the form of soft body physics. A soft body is a set of particles with constraints, but if you leave out the constraints you have a particle system. The particles are point masses though and not spheres or boxes.

**Comment by jankrassnigg** [comment]
Use raycasts and a simple bounce behavior. It's what I do in ([see here](https://ezengine.net/pages/docs/effects/particle-effects/particle-behaviors.html#raycast-behavior)) and it works really well for short lived debris. It also allows you to do performance optimizations, like only enabling the raycasts for objects close by etc, and there is no overhead for creating and managing rigid bodies.

And if you were to do GPU particles, what you would get with depth buffer collisions would be worse in quality, you just get more of them. Those effects have to be extremely short lived, otherwise the illusion falls apart.

**Comment by Ono-Sendai** [comment]
Thanks for replies guys.  
jankrassnigg: that sounds like a nice technique thanks, will try.

**Comment by Ono-Sendai** [comment]
Am trying the raycasting particle approach, seems to work quite well, if a little slowly, probably due to raycasting being expensive.  Haven't profiled yet tho.
https://twitter.com/SubstrataVr/status/1724436229700104693

**Comment by jrouwe** [comment]
What you can try is to calculate the bounding box of all particles from a single source and do ```NarrowPhaseQuery::CollectTransformedShapes```. This gives you a list of ```TransformedShape``` objects on which you can then call ```TransformedShape::CastRay```. This means you only do one broadphase query and as long as the particles are reasonably close you will get a pretty good set of objects to do narrowphase testing on. Note also that ```CastRay(const RRayCast &inRay, RayCastResult &ioHit)``` is faster than ```CastRay(const RRayCast &inRay, const RayCastSettings &inRayCastSettings, CastRayCollector &ioCollector, const ShapeFilter &inShapeFilter = { })```.

**Comment by Ono-Sendai** [comment]
Thanks Jorrit.
Apologies for prematurely besmirching the good name of Jolt performance - after profiling, most of the slowness of particles was due to overhead in my 3d engine, not ray tracing.  
Will look into your tips nonetheless.
